# RNA-seq Alignment, QC, and Quantification Nextflow Pipeline 

Bulk RNA-seq quantification pipeline with STAR and Subread FeatureCounts.

[Pipeline link](https://github.com/Meshinchi-Lab/rnaseq_count_nf)


### Exploratory Data Analysis (EDA) 

Genomic characterization of primary pediatric T-ALL samples are diagnosis and relapse. There were 51 Patient samples used in bulk illumina whole transcriptome sequencing (RNA-seq) and 8 patient samples used in Nanopore long-read whole genome sequencing (WGS) to identify gene fusions and non-coding variants leading to disease progression and relapse. 

we sequenced with the following info: - date of birth and diagnosis - the tumor specimen that was used
for sequencing (BM/PB), - immunophenotype (ETP/non-ETP and EGIL class) -
known information from cytogenetics, FISH and results from other
molecular tests, when available. - the quality of the RNA extracted (we
have RINe for these samples, calculated from TapeStation assay)


Author: Jenny Leopoldina Smith

ORCID: 0000-0003-0402-2779